src/serverobjecttype.cpp

```diff
@@ -84,25 +87,204 @@ NodeId ServerObjectType::append(
    UAPRINTLASTERROR(m_server.lastError());
-   return ret ? newNode : NodeId::Null;
+    return ret ? nodeId : NodeId::Null;
+}
+
+bool ServerObjectType::addObjectTypeFolder(const std::string& childName,
+                         const NodeId& parent,
+                         NodeId& nodeId,
+                         NodeId& requestNodeId,
+                         bool mandatory)
+{
+    NodeId newNode;
+    newNode.notNull();
+
+    if (m_server.addFolder(parent, childName, newNode, requestNodeId)) {
+        if (mandatory) {
+            return m_server.addReference(newNode,
+                                         NodeId::HasModellingRule,
+                                         ExpandedNodeId::ModellingRuleMandatory,
+                                         true);
+        }
+        if (!nodeId.isNull())
+            nodeId = newNode;
+        return true;
+    }
+    return false;
+}
+
+bool ServerObjectType::setMandatory(const NodeId& n1)
+{
+    return m_server.addReference(n1, NodeId::HasModellingRule, ExpandedNodeId::ModellingRuleMandatory, true) ==
+           UA_STATUSCODE_GOOD;
 }
 
+/*!
+    \brief addObjectTypeVariable
+    \param n
+    \param parent
+    \param nodeiD
+    \param mandatory
+    \return
+*/
+template <typename T>
+bool ServerObjectType::addObjectTypeVariable(
+    const std::string& n,
+                           const NodeId& parent,
+                           NodeId& nodeId,
+                           NodeContext* context,
+                           const NodeId& requestNodeId ,  // usually want auto generated ids
+                           bool mandatory)
+{
+    T a{};
+    Variant value(a);
+    //
+    VariableAttributes var_attr;
+    var_attr.setDefault();
+    var_attr.setDisplayName(n);
+    var_attr.setDescription(n);
+    var_attr.get().accessLevel = UA_ACCESSLEVELMASK_READ | UA_ACCESSLEVELMASK_WRITE;
+    var_attr.setValue(value);
+    var_attr.get().dataType = value.get().type->typeId;
+    //
+    QualifiedName qn(m_nameSpace, n.c_str());
+    //
+    NodeId newNode;
+    newNode.notNull();
+    //
+    if (m_server.addVariableNode(requestNodeId,
+                                 parent,
+                                 NodeId::HasComponent,
+                                 qn,
+                                 NodeId::BaseDataVariableType,
+                                 var_attr,
+                                 newNode,
+                                 context)) {
+        if (mandatory) {
+            return m_server.addReference(newNode,
+                                         NodeId::HasModellingRule,
+                                         ExpandedNodeId::ModellingRuleMandatory,
+                                         true);
+        }
+        if (!nodeId.isNull())
+            nodeId = newNode;
+        return true;
+    }
+    UAPRINTLASTERROR(m_server.lastError())
+    return false;
+}
+
+template <typename T>
+bool ServerObjectType::addHistoricalObjectTypeVariable(
```

Comment:
I don't think moving the definition of a templated function into a cpp file is a good idea.
cf: [https://raymii.org/s/snippets/Cpp_template_definitions_in_a_cpp_file_instead_of_header.html](https://raymii.org/s/snippets/Cpp_template_definitions_in_a_cpp_file_instead_of_header.html) and
[https://stackoverflow.com/questions/115703/storing-c-template-function-definitions-in-a-cpp-file](https://stackoverflow.com/questions/115703/storing-c-template-function-definitions-in-a-cpp-file)